DESIGN.md

```diff
@@ -0,0 +1,168 @@
+# Design
+
+## Implementing a Landing Zone
+
+The landing zone creation process is implemented as a [Stairway](https://github.com/DataBiosphere/stairway) flight, represented by the class `CreateLandingZoneFlight`. This flight consists of different steps one, of which is a step  
+representing a sub-flight that creates the Azure resources (`CreateLandingZoneResourcesFlight`). `CreateLandingZoneResourcesFlight` utilizes an implementation of `StepsDefinitionProvider` to define the list of steps required to create the Azure resources.
+
+Below are lists of changes required to introduce new landing zone:
+1) Implement specific steps which are responsible for Azure resource creation (see example below for `CreateVnetStep`).
+2) Create an implementation of `StepsDefinitionProvider` to define the list of Azure resource creation steps.
+3) Introduce new landing zone type by extending `StepsDefinitionFactoryType`.
+4) Update the mapping at `LandingZoneStepsDefinitionProviderFactory` based on new landing zone type. This mapping is used by the sub-flight `CreateLandingZoneResourcesFlight` to get the steps to create specific Azure resources.
+
+In the case of updating an existing landing zone, it is required to do following:
+1) Introduce new or adjusting existing step(s).
+2) In case of new steps, it is also required to include them into the corresponding implementation of `StepsDefinitionProvider`.
+
+### Landing Zone Flight Step Providers
+
+Landing zones are implemented using providers classes which return the list of necessary steps to create a landing zone. Each provider should implement the interface [StepsDefinitionProvider](https://github.com/DataBiosphere/terra-landing-zone-service/blob/main/library/src/main/java/bio/terra/landingzone/library/landingzones/definition/factories/StepsDefinitionProvider.java).
+
+### Azure Resource Flight Step
+Each step is responsible to create a certain Azure resource and is represented as a separate class. All steps are inherited from the base class named `BaseResourceCreateStep`.
+All steps are located in the following package `bio.terra.landingzone.stairway.flight.create.resource.step`. If a certain landing zone needs a new resource it is required only to introduce specific step and include it into specific provider.
+
+Below is an example of a step responsible for vnet creation. `createResource` is the central part of the class implementation and it contains the logic responsible for Azure resource creation.
+```java
+public class CreateVnetStep extends BaseResourceCreateStep {
```

Comment:
Curious to understand your thoughts for providing this code in documentation versus [providing a link directly to the source code](https://github.com/DataBiosphere/terra-landing-zone-service/blob/5607f8f968492a6691a1aa5cf8d28f50e8d7eaa7/library/src/main/java/bio/terra/landingzone/stairway/flight/create/resource/step/CreateVnetStep.java#L24) as an example?

Reply:
I have wondered that as well and questioned it on a previous iteration of this documentation. At the time it was thought that the inline code was helpful because it was somewhat of a simplification (if I remember correctly).